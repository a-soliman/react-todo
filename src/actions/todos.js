import uuid from 'uuid';

export const createTodo = 
    (
        {
            title = '', 
            note = '', 
            createdAt = 0
        } = {}
    ) => ({
        type: 'CREATE_TODO',
        todo: {
            id: uuid(),
            title,
            note,
            createdAt
        }
    }
);


export const editTodo = (id, updates) => ({
    type: 'EDIT_TODO',
    id,
    updates
});


export const removeTodo = (id) => ({
    type: 'REMOVE_TODO',
    id
});