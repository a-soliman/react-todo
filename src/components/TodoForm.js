import React from 'react';

export default class TodoForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            title: this.props.todo ? this.props.todo.title : '',
            note: this.props.todo ? this.props.todo.note : '',
            error: ''
        };
    }
    

    onTitleChange = (e) => {
        const title  = e.target.value;
        const error = '';
        this.setState(() => ({ title, error }));
    };

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    }

    onSubmit = (e) => {
        e.preventDefault();
        let error = ''
        if ( !this.state.title || this.state.title.trim().length < 1 ) {
            error = 'Please provide a title for the Todo.'
            this.setState(() => ({ error }));
        } else {
            error = ''
            this.setState(() => ({ error }));
            this.props.onSubmit({
                title: this.state.title,
                note: this.state.note
            });
        }
    };

    render() {
        return (
            <div>
                <form className="form" onSubmit={this.onSubmit}>
                    <div className="form-group">
                        <input 
                            className="form-control" 
                            value={this.state.title} 
                            onChange={this.onTitleChange}
                            type="text" 
                            placeholder="Title" 
                            autoFocus
                        />
                        {
                            this.state.error && 
                            <div className="alert alert-danger">{this.state.error}</div>
                        }
                    </div>
        
                    <div className="form-group">
                        <textarea 
                            className="form-control" 
                            value={this.state.note} 
                            onChange={this.onNoteChange}
                            placeholder="note">
                        </textarea>
                    </div>
        
                    <button type="submit">Submit</button>
                </form>
            </div>
        );
    }
}
