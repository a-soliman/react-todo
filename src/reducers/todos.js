const todosReducerDefaultState = [];

const todosReducer =( state = todosReducerDefaultState, action ) => {
    switch (action.type) {
        case 'CREATE_TODO':
            return [...state, action.todo]
        
        case 'REMOVE_TODO':
            return state.filter(todo => todo.id !== action.id)

        case 'EDIT_TODO':
            return state.map(todo => {
                if ( todo.id === action.id ) {
                    return {
                        ...todo,
                        ...action.updates
                    }
                } else {
                    return todo;
                }
            })

        default:
            return state;
    }
};

export default todosReducer;