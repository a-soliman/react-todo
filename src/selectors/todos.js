const gerVisibleTodos = (todos, { text, startDate, endDate }) => {
    return todos.filter( todo => {
        const textMatch = todo.title.toLowerCase().includes(text.toLowerCase()) || todo.note.toLowerCase().includes(text.toLowerCase());
        const startDateMatch = typeof startDate !== 'number' || todo.createdAt >= startDate;
        const endDateMatch = typeof endDate !== 'number' || todo.createdAt <= endDate;

        return textMatch && startDateMatch && endDateMatch;
    });
};

export default gerVisibleTodos;