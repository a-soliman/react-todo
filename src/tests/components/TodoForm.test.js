import React from 'react';
import { shallow } from 'enzyme';
import TodoForm from '../../components/TodoForm';
import { todos } from '../fixtures/todos';

describe('TodoForm Component', () => {
    test('Should render TodoForm correctly', () => {
        const wrapper = shallow(<TodoForm />);
        expect(wrapper).toMatchSnapshot();
    });

    test('Should render TodoForm with todo-data', () => {
        const todo = todos[1];
        const wrapper = shallow(<TodoForm todo={todo} />);
        expect(wrapper).toMatchSnapshot();
    });

    
});