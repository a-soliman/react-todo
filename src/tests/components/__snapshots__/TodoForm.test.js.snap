// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoForm Component Should render TodoForm correctly 1`] = `
<div>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <input
        autoFocus={true}
        className="form-control"
        onChange={[Function]}
        placeholder="Title"
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <textarea
        className="form-control"
        onChange={[Function]}
        placeholder="note"
        value=""
      />
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`TodoForm Component Should render TodoForm with todo-data 1`] = `
<div>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <input
        autoFocus={true}
        className="form-control"
        onChange={[Function]}
        placeholder="Title"
        type="text"
        value="todo two"
      />
    </div>
    <div
      className="form-group"
    >
      <textarea
        className="form-control"
        onChange={[Function]}
        placeholder="note"
        value="todo two note"
      />
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`TodoForm Component Should render error for invalid form submission 1`] = `
<div>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <input
        autoFocus={true}
        className="form-control"
        onChange={[Function]}
        placeholder="Title"
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <textarea
        className="form-control"
        onChange={[Function]}
        placeholder="note"
        value=""
      />
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`TodoForm Component Should render error for invalid form submission 2`] = `
<div>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <input
        autoFocus={true}
        className="form-control"
        onChange={[Function]}
        placeholder="Title"
        type="text"
        value=""
      />
      <div
        className="alert alert-danger"
      >
        Please provide a title for the Todo.
      </div>
    </div>
    <div
      className="form-group"
    >
      <textarea
        className="form-control"
        onChange={[Function]}
        placeholder="note"
        value=""
      />
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`TodoForm Component Should setState onNoteChange 1`] = `
<div>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <input
        autoFocus={true}
        className="form-control"
        onChange={[Function]}
        placeholder="Title"
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <textarea
        className="form-control"
        onChange={[Function]}
        placeholder="note"
        value="Test note"
      />
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`TodoForm Component Should setState onTitleChange 1`] = `
<div>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <input
        autoFocus={true}
        className="form-control"
        onChange={[Function]}
        placeholder="Title"
        type="text"
        value="Test Title"
      />
    </div>
    <div
      className="form-group"
    >
      <textarea
        className="form-control"
        onChange={[Function]}
        placeholder="note"
        value=""
      />
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;
